packages/app-mobile/utils/fs-driver-rn.ts

```diff
@@ -249,4 +252,57 @@ export default class FsDriverRN extends FsDriverBase {
 	public async md5File(path: string): Promise<string> {
 		throw new Error(`Not implemented: md5File(): ${path}`);
 	}
+
+	public async getExternalDirectoryPath(): Promise<string | undefined> {
+		let directory;
+		if (this.isUsingAndroidSAF()) {
+			const doc = await openDocumentTree(true);
+			if (doc?.uri) {
+				directory = doc?.uri;
+			}
+		} else {
+			directory = RNFS.ExternalDirectoryPath;
+		}
+		return directory;
+	}
+
+	public isUsingAndroidSAF() {
+		return Platform.OS === 'android' && Platform.Version > 28;
+	}
+
+	/** always returns an array */
+	public async pickDocument(options: {multiple: false}) {
```

Comment:
@jd1378, after this change the app crashes when attaching a file to a note. I had a look at it but don't know enough about Java or the RN bridge to make sense of it. Do you have any idea what can be the issue? Ref: https://discourse.joplinapp.org/t/android-pre-release-2-9-is-now-available-updated-23-october/26859/21?u=laurent

Also I'm wondering about the other changes in this pull request - have you been able to test them?

Reply:
<img width="388" alt="image" src="https://user-images.githubusercontent.com/1285584/198828511-f91426ad-43e7-404b-bfe8-f3f2b1df885c.png">

Reply:
For your information, PR #6943 might be related to this bug too.

Reply:
@laurent22 Hi. I have built 63117fd and 42a7132 on Gitpod then tested them. It's strange that both of them works fine.
<details>
<summary>Some build information</summary>

You can download these two apks for test.
https://www.filemail.com/d/fwpuenvoacukffv

For 63117fd
```
gitpod /workspace/gitpod-workspace/jd1378/packages/app-mobile/android ((63117fd...)) $ git status
HEAD detached at 63117fd
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   app/build.gradle

no changes added to commit (use "git add" and/or "git commit -a")
gitpod /workspace/gitpod-workspace/jd1378/packages/app-mobile/android ((63117fd...)) $ sha1sum app/build/outputs/apk/debug/app-debug.apk
0e217ed44f0c23c6f25d8bc264ddffe88692a2a6  app/build/outputs/apk/debug/app-debug.apk
```
For 42a7132
```
gitpod /workspace/gitpod-workspace/joplin/packages/app-mobile/android ((v2.9.11)) $ git status
HEAD detached at 42a7132
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   app/build.gradle

no changes added to commit (use "git add" and/or "git commit -a")
gitpod /workspace/gitpod-workspace/joplin/packages/app-mobile/android ((v2.9.11)) $ sha1sum app/build/outputs/apk/debug/app-debug.apk 
a239fe89f88342c775f3fadb97009a632779b055  app/build/outputs/apk/debug/app-debug.apk
```

The only change I made on both commits is this. I need this to build the apk in debug version with bundle inside and I think this doesn't affect the functionality.
```
gitpod /workspace/gitpod-workspace/joplin/packages/app-mobile/android ((v2.9.11)) $ git diff HEAD
diff --git a/packages/app-mobile/android/app/build.gradle b/packages/app-mobile/android/app/build.gradle
index cc971dd..c2ff2c5 100644
--- a/packages/app-mobile/android/app/build.gradle
+++ b/packages/app-mobile/android/app/build.gradle
@@ -78,7 +78,12 @@ import com.android.build.OutputFile
  */
 
 project.ext.react = [
-    enableHermes: false,  // clean and rebuild if changing
+    bundleInDebug: true,
+    bundleInRelease: true,
+    enableHermes: true,
+    // devDisabledInDebug: true,
+    hermesFlagsRelease: ["-O", "-output-source-map"],
+    hermesFlagsDebug: ["-O", "-output-source-map"], // plus whichever flag was required to set this away from default
 ]
 
 apply from: "../../node_modules/react-native/react.gradle"
```
</details>

As I mentioned in the [forum post](https://discourse.joplinapp.org/t/android-pre-release-2-9-is-now-available-updated-23-october/26859/21?u=wh201906), maybe you need to clean the project then build it? I can test the apks if you are willing to send them to me.

Reply:
I ran `yarn run clean` then tried to build again, and still getting the same error. I get this in logcat:

```
10-29 14:10:30.609  5206  5969 E unknown:ReactNative: Exception in native call
10-29 14:10:30.609  5206  5969 E unknown:ReactNative: com.facebook.react.bridge.NativeArgumentsParseException: SafX.openDocument got 3 arguments, expected 4
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:349)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:188)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at android.os.Handler.handleCallback(Handler.java:938)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at android.os.Handler.dispatchMessage(Handler.java:99)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at android.os.Looper.loop(Looper.java:223)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:226)
10-29 14:10:30.609  5206  5969 E unknown:ReactNative:   at java.lang.Thread.run(Thread.java:923)
```

Reply:
Make sure `gradlew clean` is being called at some point.

Reply:
You can find some files look like `index.android.bundle` in `packages/app-mobile/android/app/src/main/assets` then delete them.
Another solution is clone the repo to somewhere else then build in it.

Reply:
hi
my first guess is that safx module is not built properly, probably a gradlew clean works ?
from what I remember, we had tested that attaching files work properly last time before merging